Make queries anchorable

robertherber opened the pull request:
Resolves #50 

This will be a new major release since it changes slightly what is returned from the category and quantity queries, from simply an array of samples to:

```TypeScript
type ResponseType = {
  samples: Sample[], // category or quantity sample
  deletedSamples: DeletedSample[],
  newAnchor: string
}
```

and also adds (non-breaking) the anchor parameter as an optional input to the query:
```TypeScript
const firstRes = await queryQuantitySamples(HKQuantityTypeIdentifier.stepCount, {
  limit: 2,
})

const nextRes = await queryQuantitySamples(HKQuantityTypeIdentifier.stepCount, {
  limit: 2,
  anchor: firstRes.newAnchor,
})
```

I've given this a quick run in the example app - maybe you could have a run on it as well @ha3 and see if you have any feedback before we merge this?

ha3 commented:
@robertherber thank you for the effort! I'd been naively hoping that native side _just_ could return the _diff_ between the previous and the current call. But now I understand the API more, I see that it is impossible to do this without persisting the anchor...

So, for an app that needs to sync HealthKit data back to its own API; I guess the below approach could be used:

```ts
const anchor = storage.getItem('anchor');
let options;

if (!anchor) {
  // this is either the first time we are making the call to HealthKit or storage data is somehow wiped out. 
  options = {
    ...
  };  
} else {
  options = {
    anchor,
    to: new Date()
  };
}

const res = await queryQuantitySamples(HKQuantityTypeIdentifier.stepCount, options);

storage.setItem('anchor', res.anchor);
```

I will test it with the way I described above tomorrow and will let you know about the results. If you have any feedback regarding the approach, please let me know!

ha3 commented:
@robertherber I thought that we could compose all options with anchor, like in a paginated REST API; but if not, maybe we could throw an error if `limit` is not used along with `anchor` in dev mode?

In that case, I think your approach would be the way to go.  Only downside I can think of is that, if there is more data than the limit, there will be more trips between native and JS sides than using `to: new Date()`. But I'm not sure if it is a serious problem.

robertherber commented:
@ha3 I do believe we can compose all options. Although I'm assuming it's important that an anchor is used together with the exact same filters as specified in it's original query.

So if passing a `to` filter (`to: new Date()` in your example) I would assume it's important that the exact same date is passed when using the anchor at a later stage. But maybe I misunderstood the example.. :)

robertherber commented:
@ha3 Yeah, and I think this is both a good and bad thing. The good thing is that it seems it can be used for both paginating a query as well as fetching changes since last a query was run. I find even the Apple docs a bit unclear about the ins and outs of anchors, including what happens if an anchor is passed to a query with different parameters than the initial query.

Regarding the crash - I’m fairly sure this is how it’s been working before as well. It seems to be by design by Apple - crashing the app when requesting data you haven’t asked for access to. Although when a user denies permission it won’t crash, but instead return no data. Personally as a developer I’d prefer a warning to a crash, but I think it’s out of scope of this library to try to mitigate for this.

ha3 commented on ios/ReactNativeHealthkit.swift:
```diff
@@ -994,25 +1037,39 @@ class ReactNativeHealthkit: RCTEventEmitter {
+              "deletedSamples": deletedSamples?.map({ sample in
+                return self.serializeDeletedSample(sample: sample)
+              }) as Any,
+              "newAnchor": self.serializeAnchor(anchor: newAnchor!) as Any
```
`newAnchor` can be `nil` in some cases like when user has not given permission for the identifier. In that case serialization fails. How about `newAnchor != nil ? ... : nil`?

ha3 commented:
@robertherber sorry for the late response but we conducted some tests that I think will provide new insights.

1. About the crash, I phrased the status of the user as unauthorized; but I what I meant was that authorization prompt is presented to the user but they didn't give permission for the identifier. Previous behavior for this case was that API returns an empty array. But now it crashes. I added a comment for the concerning line that causes this.
2. Anchor seems to be corresponding to a particular object. I guess we can think of it as some hash of the object's uuid. This means that using dynamic filter options works as expected.
3. There was a problem with `limit + anchor` approach you recommended. It starts sending from the first entry but stops near end–in our case it halted at fourteen days ago. I have no idea why; but supplying `from` and `to` options solves it. Here's a working example: 

```ts
/**
  * This is important because HealthKit API seems only accepting dates initialized with a date string (i.e. 2020-06-10)
*/
const formatDate = (date) => {
  const pad = (n: number) => {
    return n < 10 ? '0' + n : n;
  };

  return (
    date.getFullYear() + '-' + pad(date.getMonth() + 1) + '-' + pad(date.getDate())
  );
};

const addDays = (date, days) => {
  date.setDate(date.getDate() + days);
  return date;
};

const getRecursiveData = async ({ limit = 10, anchor }) => {
  if (!anchor) {
    anchor = (await storage.getItem('key')) ?? undefined;
  }

  const to = new Date(formatDate(addDays(new Date(), 1)));
  const from = addDays(to, -365);

  const { deletedSamples, samples, newAnchor } = await HealthKit.queryQuantitySamples(
    HKQuantityTypeIdentifier.activeEnergyBurned,
    {
      limit,
      from,
      to,
      anchor
    }
  );

  if (samples.length > 0) {
    // persist create
  }

  if (deletedSamples.length > 0) {
    // persist delete
  }

  const hasNext = samples.length >= limit;

  if (!hasNext) {
    if (newAnchor && newAnchor !== anchor) {
      await storage.setItem('key', newAnchor);
    }
    return;
  }

  return getRecursiveData({ anchor: newAnchor, limit });
};
```

robertherber commented:
@ha3 Good catch on the nil anchors, thanks!

Regarding the dates I don't see anything odd when using non-date timestamps. What have you experienced?

I'm guessing the 2 week default limit is done with performance in mind. I've made a slight change to default to `from: new Date(0)` when a limit is specified, does that make sense and work on your data?

ha3 commented:
@robertherber we tested with the latest state of the PR; all the issues, including the one concerning timestamps are solved. As far as our tests are concerned, this PR is good to go! Thanks again!

robertherber commented:
Released in 6.0.0!
